rfc-117-containerised-runtime.md

```diff
@@ -0,0 +1,69 @@
+https://docs.google.com/document/d/1ooN7wkYhEGvceGe9Qz_HNZa-GPtrjzK_vA4vfWYVn4c/edit#heading=h.cdrr7rv9t98f) to isolate applications through containers. This will give us better control of how resources are matched to applications.
+
+The TechOps strategy is to use common components where available. GOV.UK has been following the strategy by using hosted versions of software ([ElasticSearch](https://aws.amazon.com/elasticsearch-service/) and [Postgres](https://aws.amazon.com/rds/)) and services ([Notify](https://www.notifications.service.gov.uk)) where available. Using a containerised runtime is a continuation of this policy.
+
+GOV.UK has already containerised it’s development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.
+
+### Benefits of this approach
+
+The most immediate benefit of this approach is that it will allow us to incrementally upgrade the operating system version. By moving the applications into containers, GOV.UK can enable a much lower risk upgrade path.
+
+Moving to a container based hosting environment will allow us to remove much of our legacy Puppet and Fabric code. This code is hard to maintain, and hard to hire people with the experience required to improve and upgrade. With industry and government overwhelmingly moving towards containerised platforms, it should be much easier to hire people with these skills.
+
+Containerised hosting environments can generally provide much better value for money in terms of resource usage, both because many small applications can be “bin-packed” onto the same infrastructure, and because it is much easier to autoscale applications.
+
+Running containers in production allows for a reduction in the difference between development environments and production. This can lead to bugs being found more quickly, and faster development cycles.
+
+Other improvements to GOV.UK’s development lifecycle (such as moving to continuous deployment) will be easier to implement on a container based infrastructure than they would in the current virtual machine based infrastructure.
+
+### Risks of this approach
+
+GOV.UK has not yet estimated the length of time this work will take, our current assumption is that it will take much less than 2 years to complete. There is a risk that the “unknown unknowns” GOV.UK discovers will prevent GOV.UK from finishing the containerisation work before support ends for Trusty. If this happens, GOV.UK may need to refocus our efforts to upgrade from Trusty in the existing infrastructure.
+
+GOV.UK infrastructure has been in a transitory state for several years and containerisation will prolong this state. In addition in the short term GOV.UK will be adding complexity making supporting the platform hard for RE-GOV.UK and 2ndline.
+
+Focusing on containerisation will detract from work to improve other parts of the platform.
```

Comment:
Such as?

Reply:
I think it's just that we'll need some people to be working on containerisation rather than doing other work (like migrating stuff from whitehall to content-publisher, working on the [application architecture goals](https://docs.google.com/document/d/1jEjPOFJ2s1cjQv9vHAbE-KF68LSletpnUVaG5lXlHy4/edit), doing all the other important things we need to do). We have enough people to be doing all of these things, but some things will have to have fewer people working on them if we work on this.

Maybe:

```suggestion
Focusing on containerisation will mean fewer people will be available to work on improving other parts of the platform.
```

Reply:
@richardTowers you seem to be hinting here that GOV.UK teams will be working on this? Just wondering if that is intentional or if it's not known yet whether this will be GOV.UK teams, RE or a combination of both?

Reply:
@kevindew There's no plan yet, this RFC is the start of the process. But my guess is that it'll be a combination of both.

Reply:
Relates to https://github.com/alphagov/govuk-rfcs/pull/117#discussion_r386425034 - the follow-up comment from the previous comment (oh no) - about devs having agency in these plans and the work for it.